### PR TITLE
Feature: adUnitBidLimit 

### PR DESCRIPTION
**Pull request by r-schweitzer**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
https://github.com/prebid/Prebid.js/issues/3888

Prebid.js will offer a new config setting that will cap the number of bids made per ad unit when sendAllBids is enabled. This limiter setting will be called adUnitBidLimit which will be defaulted to zero which will allow sendAllBids to function normally. When the adUnitBidLimit is set to anything above 0, e.g. 2, only the targeting for the top 2 bids will be sent to the ad server for each adunit based on CPM.

An example of the setting:

pbjs.setConfig({ adUnitBidLimit: 2 })

### CODE REVIEW

File: src/targeting.js [hunk 1]
```diff
@@ -31,16 +31,25 @@ export let filters = {
...
   let buckets = groupBy(bidsReceived, 'adUnitCode');
   // filter top bid for each bucket by bidder
   Object.keys(buckets).forEach(bucketKey => {
+    let bucketBids = [];
     let bidsByBidder = groupBy(buckets[bucketKey], 'bidderCode');
-    Object.keys(bidsByBidder).forEach(key => bids.push(bidsByBidder[key].reduce(highestCpmCallback)));
+    Object.keys(bidsByBidder).forEach(key => bucketBids.push(bidsByBidder[key].reduce(highestCpmCallback)));
+    // if adUnitBidLimit is set, pass top N number bids
+    if (adUnitBidLimit > 0) {
+      bucketBids.sort((a, b) => b.cpm - a.cpm);
```
Comment (bretg) [inline src/targeting.js]: I'd like to see the bid list randomized before sorting so that `aardvark` doesn't always win ties.

**Comment by bretg** [comment]
@mkendall07 - would like you to take a look at this one:

1) this is another way of handling how many KVPs are sent to the ad server. It's arguably superior because it truncates cleanly instead of having a partial bid. Are the methods complementary, or would this obsolete the very recent `auctionKeyMaxChars`?

2) not sure I like the proposed setConfig of `adUnitBidLimit` -- maybe `sendAllBidsLimit`?

**Comment by bretg** [comment]
Also - it's arguable -- at least worth discussing -- that deals should be given priority over bids with higher priority.

I would propose something like:

- always take the highest bid as the first block of targeting variables
- then sort on bids that contain a dealId and return the highest value deals in order
- then sort the remaining bids and them if there's room left

**Comment by bretg** [comment]
@r-schweitzer -- we're proposing the following:

1) push bid prioritization (including deals) to a different issue
2) generalize the configuration so we can work out additional controls in the (near future)

proposed config:
```
sendBidsControl: {
    bidLimit: 2
}
```

**Comment by bretg** [comment]
Ryan and I discussed in slack -- I think randomizing is the best way to go because it will resolve any question about ties and what order the bids are in now.

Apparently the list isn't in alpha-order, but it's not clear what order they're in. If they're in "first responded" order, that could be a defendable story. But if they're in "last responded" order, that would be bad. And in general, depending on the order of that list not changing might be risky.

**Comment by bretg** [comment]
Ryan found that the bid array order is based on timestamp when the bid returned.

My preference is that we still go ahead and randomize the list before picking the Top N.

1) eliminates the dependency on the order of the bids array
2) avoids any benefit from bidder ordering - as long as the response is in by the timeout, ties are random

@mkendall07 - are you good with this?

**Comment by r-schweitzer** [comment]
I just wanted to bring this back to everyone's attention. Should we be randomizing the bid array or is it fair that the bidder who responds the quickest with the higher CPM be listed first?

**Comment by bretg** [comment]
While I prefer randomizing from an algorithm perspective, I can accept the response-based ordering as long as we put a comment in somewhere to that effect. We don't really need to weigh down the browser with a sort for this edge case.

@mkendall07?

**Comment by mkendall07** [comment]
I'd prefer to keep the order as it reflects first in, first out. I'm good with keeping the scope small here and dealing with deal ids separately.

**Comment by mkendall07** [comment]
Just needs a docs PR

**Comment by bretg** [comment]
docs pr at https://github.com/prebid/prebid.github.io/pull/1516

File: src/targeting.js (hunk 1 above)
Comment (bretg) [reply src/targeting.js]: I've been told that the order of the list is chronological by time of bid response. This is ok.

**Pull request merged**